src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/LaunchProfilesDebugLaunchProvider.cs

```diff
@@ -130,34 +159,27 @@ public override async Task LaunchAsync(DebugLaunchOptions launchOptions)
                 await targetProfile.OnBeforeLaunchAsync(launchOptions, activeProfile);
             }
 
-            await DoLaunchAsync(targets.ToArray());
-
-            if (targetProfile != null)
-            {
-                await targetProfile.OnAfterLaunchAsync(launchOptions, activeProfile);
-            }
+            await DoLaunchAsync(new LaunchCompleteCallback(launchOptions, targetProfile, activeProfile), targets.ToArray());
         }
 
         /// <summary>
         /// Launches the Visual Studio debugger.
         /// </summary>
-        protected async Task<IReadOnlyList<VsDebugTargetProcessInfo>> DoLaunchAsync(params IDebugLaunchSettings[] launchSettings)
+        protected async Task DoLaunchAsync(IVsDebuggerLaunchCompletionCallback cb, params IDebugLaunchSettings[] launchSettings)
         {
             VsDebugTargetInfo4[] launchSettingsNative = launchSettings.Select(GetDebuggerStruct4).ToArray();
             if (launchSettingsNative.Length == 0)
             {
-                return Array.Empty<VsDebugTargetProcessInfo>();
+                return;
             }
 
             try
             {
                 // The debugger needs to be called on the UI thread
                 await ThreadingService.SwitchToUIThread();
 
-                IVsDebugger4 shellDebugger = await _vsDebuggerService.GetValueAsync();
-                var launchResults = new VsDebugTargetProcessInfo[launchSettingsNative.Length];
-                shellDebugger.LaunchDebugTargets4((uint)launchSettingsNative.Length, launchSettingsNative, launchResults);
-                return launchResults;
+                IVsDebuggerLaunchAsync shellDebugger = await _vsDebuggerService.GetValueAsync();
+                shellDebugger.LaunchDebugTargetsAsync((uint)launchSettingsNative.Length, launchSettingsNative, cb);
```

Comment:
We call `OnBeforeLaunchAsync()` and `OnAfterLaunchAsync()` even if Launch fails. In this new implementation, `OnAfterLaunchAsync()` is executed only if Launch succeeds. The change makes sense for me, but not sure if `OnAfterLaunchAsync()` used for other purposes other than just call it after some work.